service/worker/scanner/executions/scavenger.go

```diff
@@ -175,23 +186,51 @@ func (s *Scavenger) awaitExecutor() {
 	}
 }
 
-func (s *Scavenger) emitStats() {
-	// TODO: implement this, this will emit metrics after a full run of executor scavenger is finished
-}
-
-// newTask returns a new instance of an executable task which will process a single execution
-func (s *Scavenger) newTask(namespaceID, workflowID, runID string) executor.Task {
+// newTask returns a new instance of an executable task which will process a single mutableState
+func (s *Scavenger) newTask(shardID int32, rateLimiter quotas.RateLimiter) executor.Task {
 	return &executorTask{
-		executionKey: executionKey{
-			namespaceID: namespaceID,
-			workflowID:  workflowID,
-			runID:       runID,
-		},
-		scvg: s,
+		shardID:     shardID,
+		rateLimiter: rateLimiter,
+		scvg:        s,
 	}
 }
 
 // Run runs the task
 func (t *executorTask) Run() executor.TaskStatus {
-	return t.scvg.validateHandler(&t.executionKey)
+	return t.scvg.validateShardHandler(t.shardID, t.rateLimiter)
+}
+
+func (s *Scavenger) validateShardHandler(shardID int32, rateLimiter quotas.RateLimiter) executor.TaskStatus {
+	validationResult := s.shardValidatorFactory(rateLimiter).validate(shardID)
+	dumpShardValidationReport(validationResult, s.logger, s.metrics)
+	return executor.TaskStatusDone
+}
+
+func dumpShardValidationReport(report *ShardValidationResult, logger log.Logger, metricClient metrics.Client) {
+	metricsScope := metricClient.Scope(metrics.ExecutionsScavengerScope)
+	metricsScope.AddCounter(metrics.ScavengerDBRequestsCount, report.TotalDBRequests)
+	metricsScope.UpdateGauge(
+		metrics.ExecutionsScavengerExecutionsCount,
+		float64(report.ScanStats.TotalExecutionsCount))
+	metricsScope.UpdateGauge(
+		metrics.ExecutionsScavengerCorruptedExecutionsCount,
+		float64(report.ScanStats.CorruptedExecutionsCount))
+
+	if !report.IsFailure {
+		return
+	}
+	for _, executionFailure := range report.ExecutionFailures {
+		for validator, validatorFailure := range executionFailure.Failures {
+			failureScope := metricsScope.Tagged(
+				metrics.ValidatorTag(validator),
+				metrics.FailureTag(validatorFailure.failureReasonTag),
+			)
+			failureScope.IncCounter(metrics.ScavengerValidationFailuresCount)
+		}
+		logger.Warn(
+			"Validation failed for executions shard.",
+			tag.ShardID(report.ShardID),
+			tag.Error(report.Error),
+			tag.Value(executionFailure))
+	}
```

Comment:
this error message does not seem to provide enough information?

Reply:
do not just dump the execution failure, log the namespace / workflow / run as part of the dedicated tag?